fem/dfem/doperator.hpp

```diff
@@ -908,6 +941,122 @@ void DifferentiableOperator::AddIntegrator(
             });
          }
 
+         // First Input index of the derivative
+         const size_t d_input_idx = [d_field_idx, &input_to_field]
+         {
+            for (size_t i = 0; i < input_to_field.size(); i++)
+            {
+               if (input_to_field[i] == d_field_idx)
+               {
+                  return i;
+               }
+            }
+            return size_t(SIZE_MAX);
+         }();
+
+         const int trial_vdim = GetVDim(fields[d_field_idx]);
+         const int num_trial_dof =
+            get_restriction<entity_t>(fields[d_field_idx], element_dof_ordering)->Height() /
+            inputs_vdim[d_input_idx] / num_entities;
+         const int num_trial_dof_1d =
+            input_dtq_maps[d_input_idx].B.GetShape()[DofToQuadMap::Index::DOF];
+
+         Vector Ae_mem(num_test_dof * test_vdim * num_trial_dof * trial_vdim *
+                       num_entities);
+         Ae_mem = 0.0;
+
+         // Quadrature point local derivative cache for each element, with data
+         // layout:
+         // [test_vdim, test_op_dim, trial_vdim, trial_op_dim, qp, num_entities].
+         derivative_qp_caches[derivative_id] = Vector(test_vdim * test_op_dim *
+                                                      trial_vdim *
+                                                      total_trial_op_dim * num_qp * num_entities);
+         // Create local references for MSVC lambda capture compatibility
+         auto& fields_ref = this->fields;
+         auto& derivative_qp_caches_ref = this->derivative_qp_caches[derivative_id];
+
+         derivative_setup_callbacks[derivative_id].push_back(
```

Comment:
The new 'cache' concept is powerful, and would benefit from some explanations.
Documenting the underlying ideas in the PR description, as well as through inline comments in the code, would make it easier to fully appreciate it and follow the computations.